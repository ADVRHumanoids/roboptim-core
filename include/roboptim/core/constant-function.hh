// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_CONSTANT_FUNCTION_HH
# define ROBOPTIM_CORE_CONSTANT_FUNCTION_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <roboptim/core/linear-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_function
  /// @{

  /// \brief Constant function.
  ///
  /// Implement a constant function using the formula:
  /// \f[f(x) = offset\f]
  /// where \f$offset\f$ is set when the class is instantiated.
  class ROBOPTIM_DLLAPI ConstantFunction : public LinearFunction
  {
  public:
    /// \brief Build an constant function.
    ///
    /// \param offset constant function offset
    ConstantFunction (const vector_t& offset) throw ();
    ~ConstantFunction () throw ();

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    void impl_compute (result_t& , const vector_t&) const throw ();
    void impl_gradient (gradient_t&, const vector_t&, size_type = 0)
      const throw ();
    void impl_jacobian (jacobian_t&, const vector_t&) const throw ();

  private:
    const vector_t offset_;
  };

  /// Example shows constant function use.
  /// \example constant-function.cc

  /// @}

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_CONSTANT_FUNCTION_HH
