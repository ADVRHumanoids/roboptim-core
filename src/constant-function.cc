// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include <boost/numeric/ublas/io.hpp>
#include <boost/numeric/ublas/matrix.hpp>
#include <boost/numeric/ublas/vector.hpp>

#include <roboptim/core/indent.hh>
#include <roboptim/core/constant-function.hh>

namespace roboptim
{
  ConstantFunction::ConstantFunction (const vector_t& offset)
    throw ()
    : LinearFunction (offset.size (), offset.size (), "constant function"),
      offset_ (offset)
  {
  }

  ConstantFunction::~ConstantFunction () throw ()
  {
  }

  void
  ConstantFunction::impl_compute (result_t& result,
				  const argument_t&)
    const throw ()
  {
    result = this->offset_;
  }

  void
  ConstantFunction::impl_jacobian (jacobian_t& jacobian,
				   const argument_t&) const throw ()
  {
    jacobian.clear ();
  }

  void
  ConstantFunction::impl_gradient (gradient_t& gradient,
				   const argument_t&,
				   size_type) const throw ()
  {
    gradient.clear ();
  }

  std::ostream&
  ConstantFunction::print (std::ostream& o) const throw ()
  {
    return o << "Constant function" << incindent << iendl
             << "offset = " << this->offset_ << iendl
             << decindent;
  }

} // end of namespace roboptim
