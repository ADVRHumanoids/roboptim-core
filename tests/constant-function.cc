// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>
#include <boost/numeric/ublas/io.hpp>

#include "common.hh"
#include <roboptim/core/constant-function.hh>

using namespace roboptim;

int run_test ()
{
  ConstantFunction::vector_t offset (4);
  offset[0] = 12.;
  offset[1] = 46.;
  offset[2] = 2.;
  offset[3] = -9.;

  ConstantFunction cst (offset);

  ConstantFunction::vector_t x (4);
  x.clear ();

  std::cout
    << cst << std::endl
    << "Evaluate: " << std::endl
    << cst (x) << std::endl
    << "Gradient: " << std::endl
    << cst.gradient (x) << std::endl
    << "Jacobian: " << std::endl
    << cst.jacobian (x) << std::endl;

  return 0;
}

GENERATE_TEST ()
