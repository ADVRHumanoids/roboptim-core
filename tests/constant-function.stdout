Constant function
  offset = [4](12,46,2,-9)
  
Evaluate: 
[4](12,46,2,-9)
Gradient: 
[4](0,0,0,0)
Jacobian: 
[4,4]((0,0,0,0),(0,0,0,0),(0,0,0,0),(0,0,0,0))
